Create a full Bot using Twitter Streaming API

leonplondon opened the pull request:
# Summary
Create bot using twits filtered from Twitter Stream API.

# Trello Card
https://trello.com/c/0eY87hyO/10-crear-un-bot

leonplondon committed:
Create a full Bot using Twitter Streaming API

yanzacd commented on src/main/java/wolox/reactortraining/services/BotService.java:
```diff
@@ -38,4 +66,76 @@
+            .map(Tweet::getText)
+            .take(500)
+            .filter(text -> {
+                for (String topic : topics) {
```
You should do a refactor on this, another method with descriptive name for claritier

yanzacd commented on src/main/java/wolox/reactortraining/services/BotService.java:
```diff
@@ -38,4 +66,76 @@
+            .collectList()
+            .map(stringList -> String.join(" ", stringList))
+            .map(text -> {
+                BotDto botDto = new BotDto();
```
the same before comment

yanzacd commented on src/main/java/wolox/reactortraining/services/BotService.java:
```diff
@@ -38,4 +66,76 @@
+                Mono.zip(createTopics(user, botCreationDto.getTopics()), Mono.just(user)))
+            .flatMap(tuple -> {
+                User user = tuple.getT2();
+                List<Topic> topics = tuple.getT1();
```
refactor

yanzacd commented on src/main/java/wolox/reactortraining/services/TwitterService.java:
```diff
@@ -16,4 +31,38 @@ public SearchResults searchSpringBoot(String query) {
+        ConnectableFlux<Tweet> tweetsListener = Flux
+            .create((FluxSink<Tweet> emitter) -> {
+                StreamListener streamListener = new StreamListenerDelegate(emitter);
+                List<StreamListener> listeners = Collections.singletonList(streamListener);
```
refactor

yanzacd commented on src/main/java/wolox/reactortraining/exceptions/MatchingTwitsNotFound.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(code = HttpStatus.NOT_FOUND, reason = "Twits about topics not found in stream")
```
tweets :)

Jhovannycanaspino commented on src/main/java/wolox/reactortraining/controllers/BotController.java:
```diff
@@ -39,4 +40,11 @@
     }
+
+    @PostMapping("/create")
+    public Mono<Void> createBot(@Valid @RequestBody BotCreationDto botCreationDto) {
```
place the http response in methods that return void

Jhovannycanaspino commented on src/main/java/wolox/reactortraining/services/BotService.java:
```diff
@@ -38,4 +66,76 @@
+        return Flux
+            .fromIterable(topics)
+            .map(description -> {
+                TopicDto topicDto = new TopicDto();
```
this can be simplified: TopicDto::new

leonplondon replied:
It can't be because we are invoking another object's method

Jhovannycanaspino commented on src/main/java/wolox/reactortraining/services/BotService.java:
```diff
@@ -38,4 +66,76 @@
+                return topicDto;
+            })
+            .map(dto -> {
+                Topic topic = new Topic(dto);
```
same

leonplondon replied:
Same

leonplondon committed:
Fixed PR#5

leonplondon committed:
Added annotated status code to create bot endpoint

yanzacd commented on src/main/java/wolox/reactortraining/services/BotService.java:
```diff
@@ -16,6 +32,50 @@
+    private static boolean tweetHasTopics(String tweet, List<String> topics) {
+        boolean contains = false;
+
+        for (String topic : topics) {
```
you should use operations lambda, don't use for, use a foreach

leonplondon replied:
The forEach does not allow me to return the value of contains to the upper function `tweetHasTopics` without turning the variable `contains` an Atomic reference. Should I replace that with this last alternative?

yanzacd commented on src/main/java/wolox/reactortraining/models/Topic.java:
```diff
@@ -29,4 +32,18 @@ public Topic() {
     }
+
+    public void addUser(User user) {
+        if (users == null) {
```
refactor this in global var

yanzacd commented on src/main/java/wolox/reactortraining/models/User.java:
```diff
@@ -29,5 +32,13 @@ public User() {
     }
+
+    public void addTopics(List<Topic> topics) {
+        if (this.topics == null) {
```
refactor this in global var

yanzacd commented on src/main/java/wolox/reactortraining/models/Topic.java:
```diff
@@ -29,4 +32,18 @@ public Topic() {
+        if (!users.contains(user)) {
+            users.add(user);
+        } else {
+            int index = users.indexOf(user);
```
skip this

yanzacd commented on src/main/java/wolox/reactortraining/services/BotService.java:
```diff
@@ -38,4 +99,56 @@
+        return twitterService
+            .getTweetsStreamPipe()
+            .map(Tweet::getText)
+            .take(500)
```
put 500 value in constant or environment var, how you prefer

leonplondon replied:
I gonna add it to `twitter.properties`

leonplondon committed:
Fixed create bot.

Removed @ResponseStatus becuase it avoids subscription to returned publishers.
Fixed methods to relate User & Topic models.
Added lazy @DBRef to models to avoid infinite recursion when updating 'relationships'
